### PR TITLE
feat: actually tell users when verified addresses dont match

### LINKED ISSUE #6814: Handle multiple Ledgers and account management
### Overview

The ledger device ID is the same for every device, meaning query cache using wallet `deviceId` will be invalid if a different ledger is connected.

### References and additional details

Originally noted here :
https://github.com/shapeshift/web/pull/6780#discussion_r1588979619

### Acceptance Criteria

Account management operates as intended across different ledger devices for all states and user flows.

SPIKE:

1. can we auto open (and close / re-open) apps on ledger devices
2. if so, this can be implemented in both connection / pairing flow and also to confirm device identity
3. if not, lazy check of device identity on next wallet interaction (confirming account 0 matches expected account 0). If doesn't match prompt user. If match no action, proceed. 

research #1 and create needed follow up tickets for 2 and 3 depending on out come. 

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

If address verification failed on ledger wallet (due to device mismatch or other reason), there is currently no notification of this to the user. This PR aims to tell the user when addresses dont match and prompt them to connect the correct device and try again.

This PR is focused on the trade page only, and follow-ups will be required to ensure the same fixes are applied to other pages.

## Issue (if applicable)

progresses #6814

## Risk

> High Risk PRs Require 2 approvals

High risk as this affects a protection against sending funds to the wrong address.

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

Ledger wallet, all protocols, trades.

## Testing

- Check non-ledger wallets can trade as expected.
- Ledger trades work on all chains as expected in the happy common case.
- Users cannot trade when intentionally connecting the wrong ledger device. Note users will be able to use the app, the protection is at the time of signing.
- When the correct device is connected, address verification succeeds.

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

Using the correct device to verify address, happy case where user verifies without canceling:

https://github.com/user-attachments/assets/659a2d99-2c1d-4038-bb41-a270c9fb654f

Using the correct device to verify address, case where user rejects 2 times and then approves:

https://github.com/user-attachments/assets/8c3c569e-116b-440f-8a74-8659beabca06

Opened app with wrong device, user tries to verify addresses and is protected by the mechanism:

https://github.com/user-attachments/assets/aef9af54-1659-4f57-ac2a-e41461b7496e

Trying REALLY hard to break it:
1. commence a trade with correct device
2. verify AVAX address -> OK
3. connect a different device
4. verify RUNE address -> ERROR
5. connect the correct device
6. verify RUNE address -> OK

https://github.com/user-attachments/assets/ac851664-2fff-42cd-a6cf-57304ca5ab93

### CODE REVIEW

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 1]
```diff
@@ -163,13 +158,64 @@ export const VerifyAddresses = () => {
...
+  const clearErrorMessage = useCallback((type: 'sell' | 'buy') => {
+    if (type === 'sell') {
+      setSellErrorMessage(undefined)
+    } else if (type === 'buy') {
+      setBuyErrorMessage(undefined)
+    }
+  }, [])
+
+  const setStatus = useCallback(
+    (type: 'sell' | 'buy', status: 'pending' | 'verified' | 'error' | 'notRequired') => {
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:194]: Can we make `status` an enum type and then use it throughout this file? E.g. `TradeStatus.Pending` etc

Ideally `type` would be too, though not a specific concern of this PR.

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 2]
```diff
@@ -163,13 +158,64 @@ export const VerifyAddresses = () => {
...
+      // if the sell and buy addresses are the same, we can skip the verification for the other address
+      console.log('sellAddress', sellAddress)
+      console.log('buyAddress', buyAddress)
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:197]: rm?

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 3]
```diff
@@ -50,7 +50,10 @@ export const VerifyAddresses = () => {
   const [isSellVerifying, setIsSellVerifying] = useState(false)
   const [isBuyVerifying, setIsBuyVerifying] = useState(false)
 
-  const [verifiedAddresses, setVerifiedAddresses] = useState(new Set<string>())
+  const [buyStatus, setBuyStatus] = useState('pending')
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:53]:
```suggestion
  const [buyAddressVerificationStatus, setBuyAddressVerificationStatus] = useState('pending')
```

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 4]
```diff
@@ -50,7 +50,10 @@ export const VerifyAddresses = () => {
...
+  const [sellStatus, setSellStatus] = useState('pending')
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:54]:
```suggestion
  const [sellAddressVerificationStatus, setSellAddressVerificationStatus] = useState('pending')
```

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 5]
```diff
@@ -50,7 +50,10 @@ export const VerifyAddresses = () => {
...
+  const [buyErrorMessage, setBuyErrorMessage] = useState<string | undefined>()
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:55]:
```suggestion
  const [buyAddressVerificationErrorMessage, setBuyAddressVerificationErrorMessage] = useState<string | undefined>()
```

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx [hunk 6]
```diff
@@ -50,7 +50,10 @@ export const VerifyAddresses = () => {
...
+  const [sellErrorMessage, setSellErrorMessage] = useState<string | undefined>()
```
Comment (0xApotheosis) [inline src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx:56]:
```suggestion
  const [sellAddressVerificationErrorMessage, setSellAddressVerificationErrorMessage] = useState<string | undefined>()
```

**Review by 0xApotheosis — COMMENTED** [review]
Code looks good, added some thoughts on naming.

I don't have a second Ledger to do a functional pass on this PR.

**Review by gomesalexandre — COMMENTED** [review]
Code conceptually looks sane at a high-level (minus naming issues spotted by @0xApotheosis), confirmed this does what it says on the box at runtime!

Happy to stamp once review comments have been tackled.

### Ledger connected with correct addies for the swap

- Same buy/sell addies - still automagically verifies the second

image
image
image

- Cross-chain - still prompts for two confirmations

image
image
image

### Ledger connected with invalid addies for the swap

image

**Review by 0xApotheosis — APPROVED** [review]
✅ Matching, different addresses

Screenshot 2024-09-04 at 10 49 41 AM

✅ Matching, same addresses shortcuts to only need on validation

Screenshot 2024-09-04 at 10 51 08 AM

✅ Correctly alerts on address mismatch

Screenshot 2024-09-04 at 10 59 09 AM

@gomesalexandre I'm happy with this if you are!

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 1 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/ffcadbf199a3b8f15fcbe5e8199b5fb1985c2fe4

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 2 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/bfd7cb63f74c418c4ca4c06e9a10a4f82a1bb0d3

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 3 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/bfd7cb63f74c418c4ca4c06e9a10a4f82a1bb0d3

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 5 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/bfd7cb63f74c418c4ca4c06e9a10a4f82a1bb0d3

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 4 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/bfd7cb63f74c418c4ca4c06e9a10a4f82a1bb0d3

File: src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx (hunk 6 above)
Comment (gomesalexandre) [reply src/components/MultiHopTrade/components/VerifyAddresses/VerifyAddresses.tsx]: https://github.com/shapeshift/web/pull/7639/commits/bfd7cb63f74c418c4ca4c06e9a10a4f82a1bb0d3

**Review by gomesalexandre — APPROVED** [review]
Diff LGTM now! Get in.

**Pull request merged**